sharding/node/backend.go

```diff
@@ -0,0 +1,193 @@
+	cli "gopkg.in/urfave/cli.v1"
+)
+
+// ShardEthereum is a service that is registered and started when geth is launched.
+// it contains APIs and fields that handle the different components of the sharded
+// Ethereum network.
+type ShardEthereum struct {
+	shardConfig  *params.ShardConfig  // Holds necessary information to configure shards.
+	txPool       *txpool.ShardTXPool  // Defines the sharding-specific txpool. To be designed.
+	actor        sharding.Actor       // Either notary, proposer, or observer.
+	shardChainDb ethdb.Database       // Access to the persistent db to store shard data.
+	eventFeed    *event.Feed          // Used to enable P2P related interactions via different sharding actors.
+	smcClient    *mainchain.SMCClient // Provides bindings to the SMC deployed on the Ethereum mainchain.
+
+	// Lifecycle and service stores.
+	services map[reflect.Type]sharding.Service // Service registry.
+	lock     sync.RWMutex
+}
+
+// New creates a new sharding-enabled Ethereum instance. This is called in the main
+// geth sharding entrypoint.
+func New(ctx *cli.Context) (*ShardEthereum, error) {
+
+	shardEthereum := &ShardEthereum{
+		services: make(map[reflect.Type]sharding.Service),
+	}
+
+	path := node.DefaultDataDir()
+	if ctx.GlobalIsSet(utils.DataDirFlag.Name) {
+		path = ctx.GlobalString(utils.DataDirFlag.Name)
+	}
+
+	endpoint := ctx.Args().First()
+	if endpoint == "" {
+		endpoint = fmt.Sprintf("%s/%s.ipc", path, mainchain.ClientIdentifier)
+	}
+	if ctx.GlobalIsSet(utils.IPCPathFlag.Name) {
+		endpoint = ctx.GlobalString(utils.IPCPathFlag.Name)
+	}
+
+	passwordFile := ctx.GlobalString(utils.PasswordFileFlag.Name)
+	depositFlag := ctx.GlobalBool(utils.DepositFlag.Name)
+	actorFlag := ctx.GlobalString(utils.ActorFlag.Name)
+
+	smcClient, err := mainchain.NewSMCClient(endpoint, path, depositFlag, passwordFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// Adds the initialized SMCClient to the ShardEthereum instance.
+	shardEthereum.smcClient = smcClient
+
+	if err := shardEthereum.registerP2P(); err != nil {
+		return nil, err
+	}
+
+	if err := shardEthereum.registerTXPool(actorFlag); err != nil {
+		return nil, err
+	}
+
+	if err := shardEthereum.registerActorService(actorFlag); err != nil {
+		return nil, err
+	}
+
+	return shardEthereum, nil
+}
+
+// Start the ShardEthereum service and kicks off the p2p and actor's main loop.
+func (s *ShardEthereum) Start() {
+	log.Info("Starting sharding node")
+	for kind, service := range s.services {
+		// Start the next service.
+		if err := service.Start(); err != nil {
+			log.Error(fmt.Sprintf("Could not start service: %v, %v", kind, err))
+			s.Close()
+		}
+	}
+}
+
+// Close handles graceful shutdown of the system.
+func (s *ShardEthereum) Close() {
+	for kind, service := range s.services {
+		if err := service.Stop(); err != nil {
+			log.Crit(fmt.Sprintf("Could not stop the following service: %v, %v", kind, err))
+		}
+	}
+	log.Crit("Stopping sharding node")
```

Comment:
This should be Info level. The error that a service could not be stopped is critical to the user while this is statement here is more of a convenience log.